tests/test_search.py

```diff
@@ -176,7 +179,10 @@ def test_IndexBuilder():
         'index': {'docname1_1', 'docname1_2', 'docname2_1', 'docname2_2'},
         'test': {'docname1_1', 'docname1_2', 'docname2_1', 'docname2_2'},
     }
-    assert index._title_mapping == {'section_titl': {'docname1_1', 'docname1_2', 'docname2_1', 'docname2_2'}}
+    assert index._title_mapping == {
```

Comment:
Here, I see that the values are sets and thus can be unordered. Is it possible to actually have a more reliable guarantee on the `title_mapping` actually?

Reply:
To check that I understand correctly: use an assertion that checks not only the _contents_ of the title mapping, but also the _ordering_ of the elements inside it?

(makes sense to me, order can be significant for index data)

Reply:
If my understanding there was correct: I'd prefer to handle that in a separate issue thread and PR, if that's OK.  Or, if I misunderstood: let's make sure to resolve that first.

Reply:
If there is more than one occurrence of that, yes we could discuss it in another thread.

Reply:
Yep; I think it would be applicable for both `_title_mapping` and `_mapping`; because both of those currently use a pattern of `<mapping>.setdefault(..., set()).add(...)` (implying that insertion-order may not match subsequent access iteration order):

https://github.com/sphinx-doc/sphinx/blob/b70578f80c78b00c6571203f3204126742367bce/sphinx/search/__init__.py#L447-L464

Reply:
Correction: any _runtime_ (in-memory) differences in the ordering of objects within the `_title_mapping` and `_mapping` set entries are later resolved into deterministic order by an [explicit sort](https://github.com/sphinx-doc/sphinx/blob/b70578f80c78b00c6571203f3204126742367bce/sphinx/search/__init__.py#L380) that occurs during [freezing](https://github.com/sphinx-doc/sphinx/blob/b70578f80c78b00c6571203f3204126742367bce/sphinx/search/__init__.py#L388) (serialization) of the search index (thanks to d24bd73d0ce8071459a1056691a1934e1ca12194).

I'm not aware of any other current bugs causing nondeterminism in our `searchindex.js` output since v7.3.0 (95fb0e5e570481cfb5e98990f657e3ef0fc26adc included in that release handled a similar situation).

Reply:
So `_title_mapping` are unsorted in the tests, but in production they are sorted? if so, a comment is needed (at least to remember the reason).

Reply:
Yep, pretty much.  I'll be a bit verbose: 

* In-memory (and therefore in the tests), each value in `_title_mapping` may store elements in arbitrary ordering.
* The tests use Python `set` equality comparison, and that is order-agnostic (as it should be) -- so the tests aren't really checking the order, even though it might appear from the code that they would.
* The output of a Sphinx production build to output a `searchindex.js` sorts these entries.

What you suggest about a comment makes sense...

...however.. perhaps I could attempt a separate PR to refactor the code so that sorting occurs in-place and in-memory, and to adjust the unit tests to assert on that.  Those would provide stronger properties than a comment in the code.

Reply:
No need for that. It's better that at runtime, the objects are represented in sets since they are more efficient. It's only for serialization that you need to sort to ensure reproducibility.

Reply:
Ok, agreed (storing the lists in-place for ordered iteration would probably add various types of overhead).

I've pushed a commit to add some commentary as a docstring on the relevant method and also in the unit tests.